README.md
# cs5293sp22-project1
### Author : Pradipkumar Rajasekaran

__Summary:__

The goal of this project is to redact data from text files. The text files contain Emails from the Enron scandal. The text file is given as input,
which is cleaned and taken as a string. This string then goes through several other functions as specified by the user in the command line arguments.
Each function identifies the substring to redact and does the redaction by overwrting with a block. The redacted files are outputed with the .redacted extension.
 
## Developement Process

- Created redacted.py with the function definifions and function calls in the main loop.
- The file names are taken as input and data from multiple files are sored with a seperator to distinguish the files.
- The functons detect the substrings to redact using spacy, nltk or regex.
- The redacted output is stored in the path given by the user.

__Installation__


1. Clone the repository- git clone https://github.com/CurSpace/cs5293sp22-project1.git
2. Navigate to the project0 folder - cd "cs5293sp22-project1/project1"
3. Install pipenv - pip install pipenv
4. Install the required packages - pipenv install -r requirements.txt
5. Navigate to test folder and run pytest - pipenv run pytest testing_redactor.py


__Python packages used:__

- argparser
- spacy
- nltk
- pytest
- sys
- glob

### Description

- The redactor.py controls the flow of operations.
- The functions are implemted in the redactor.py
- Run the program by:
```
   pipenv run python redactor.py --input [filename] --names --dates --names --phones --genders --address --concept [CONCEPT] --output --stats stderr
```
- the input can be one or more text files
   Ex: 
   
   ```
       pipenv run python redactor.py --input 5.txt 6.txt 7.txt --names --phones --genders --address --concept email --output file --stats stderr
   ```

 
 __Description of User Defined Functions:__
 
 1. input_files(files_in) - takes all the text files and stores them as a single string  with filesep as the seperator between the files.
 
 2. redact_all_names(data) - tokenizes the string using spacy and redacts all names with label PERSON

 3. redact_all_dates(data) - redacts dates using regex

 4. redact_all_phonembers(data) - redacts phone numbers using regex

 5. redact_all_genders(data) - redacts genders uisng a gender list. If the word that incicates gender is present in the gender list, then 
    the gender indicator is redacted.

 6. redact_all_addressess(data) - redacts addressess using regex

 7. redact_all_concepts(data,concept) - redacts entire sentence containing synonym using wordnet

 8. redacted_output(files_in,redacted_file) - stores the redacted text into it's corresponding .redacted file

 9. get_stats(type_of,count) - collects the number of redactions of names,dates,phones,genders,addressess and concepts from the respective 
                        	functioins and appends the result to a global list

 10. write_stats(path,store_stats) - stores the global list returned by write_stats() into the file specified by the user.
    
  
        

    __Sample Output:__
   
 ```
5.redacted

['["Message-ID: <████████.█████████████.JavaMail.evans@thyme> Date: Sun, 25 Nov 2001 20:44:49 -0800 (PST) From: enerfaxdaily@enerfax.com \n███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████ Slow browser may need to wait 15 seconds for your email to appear, then scroll to view. mailto:leave-enerfaxdaily-794030K@relay.netatlantic.com to unsubscribe enerfaxdaily as: fermis@ect.enron.com forward to web@enerfax.com  If your email does not support this web version, please subscribe to the free text version by writing subscribe@enerfax.com <mailto:subscribe@enerfax.com> Please visit our website to view today\'s gas and power prices and news by clicking <http://www.enerfax.com>   If you received this in error or no longer wish to subscribe click this link mailto:leave-enerfaxdaily-794030K@relay.netatlantic.com or forward to web@enerfax.com ███ ███ ███ ████  ", \'

6.redacted

['["Message-ID: <████████.█████████████.JavaMail.evans@thyme> Date: Sun, 25 Nov 2001 20:44:49 -0800 (PST) From: enerfaxdaily@enerfax.com \n███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████ Slow browser may need to wait 15 seconds for your email to appear, then scroll to view. mailto:leave-enerfaxdaily-794030K@relay.netatlantic.com to unsubscribe enerfaxdaily as: fermis@ect.enron.com forward to web@enerfax.com  If your email does not support this web version, please subscribe to the free text version by writing subscribe@enerfax.com <mailto:subscribe@enerfax.com> Please visit our website to view today\'s gas and power prices and news by clicking <http://www.enerfax.com>   If you received this in error or no longer wish to subscribe click this link mailto:leave-enerfaxdaily-794030K@relay.netatlantic.com or forward to web@enerfax.com ███ ███ ███ ████  ", \'

7.redacted

\', \'Message-ID: <███████.█████████████.JavaMail.evans@thyme> Date: Sat, 26 Nov 0001 00:06:35 -0800 (PST) From: ipayit@enron.com To: frank.ermis@enron.com Subject: Action Requested:  Past Due Invoice for User: FRANK ERMIS Mime-Version: 1.0 Content-Type: text/plain; charset=us-ascii Content-Transfer-Encoding: 7bit X-From: iPayit </O=ENRON/OU=NA/CN=RECIPIENTS/CN=MBX IPAYIT> X-To: Ermis, █████ </O=ENRON/OU=NA/CN=RECIPIENTS/CN=FERMIS> X-cc:  X-bcc:  X-Folder:  ███████████████████ 1 Ermis, ████████████████████: Ermis-F X-FileName: fermis (Non-Privileged).pst  Alert!   You are receiving this message because you have an unresolved invoice in your iPayit in-box that is past due.  It is critical that you login to iPayit and take immediate action to resolve this invoice.   Remember, you play an important role in ensuring that we pay our vendors on time.   Tip!:  You must login to the system to forward this invoice to another user.   To launch iPayit, click on the link below: http://iPayit.enron.com Note:  Your iPayit User ID and Password are your eHRonline/SAP Personnel ID and Password.  First time iPayit user?  For training materials, click on the link below: http://isc.enron.com/site/doclibrary/user/default.asp  Need help? North America: ISC Call Center at ██████████████. Europe: European Accounts Payable at +44-20-7783-7520. \n██████████████████████████████████████████████████████████████████████']

stderr

Names:7
Phones:7
Genders:4
Concepts:3

Note: stats are not mentioned if there are no occurrences.
 ```
__Assumptions__

1. Input files are text files.

2. General regex patterns will be able to find required items.

3. Spacy will find all names with label PERSON.

4. wordnet will accurately redact sentences with specified concept synonyms.

5. Names do not have to be redacted if they appear in emails. 

6. Phonenumbers are of format (713) 345-472.

7. Gender reference are limited to the words mentioned in the gender list.

8. Addressess are of format \d{1,6}\s(?:[A-Za-z0-9#]+\s){0,7}(?:[A-Za-z0-9#]+,)\s*(?:[A-Za-z]+\s){0,3}(?:[A-Za-z]+,)\s*[A-Z]{2}\s*\d{5}

9. wordnet is sufficient to identify synonyms.

10. Output is stored in the locaton specifiec in the user specified locaton with .redacted extension.

11. Stats are stored in the current working directory with the user specifiec file name.
 
__BUGS__

1. The presence of \n interferes with detecting certain ideas like gender. Therefore replacing \n with empty space.

2. Spacy is unable to detect uncommon names like Emris.

3. Redacts wrong 2 numbers at the start of every email ['5941235', '1075855165188', '(713) 345-4727']. CommonRegex is unable to distinguish email identifier and phonenumber.

4. Phones of the format +61-2-9229-2336 are not matched as they are not common north american formats.

5. Redacting concepts interferes with the other options to redact because they have been already redacted, this affects the stats and not all redactions are collected in the stats file.

   


__Testing__

1. test_datainput()

   -> tests if data exists and is not empty

2. test_phonenumbers.py

   -> tests if the lenght of data returned by redact_all_phone_numbers() is greater than zero and makes sure the type of data is not None.

3. test_genders() 

   -> checks if the blocks and the number of redactions are same
   
4. test_names() 
     -> tests if the numbers of blacks is the same as teh number of matched patterns 

5. test_genders() - 
 
    -> checks if the blocks and the number of patterens matched are the same.

 
 



CHANGES.md
# Changes
* Moderate amount of Features Found - Gender --- Improved genders by using a better gender list and ingnoring the characters if they are a substring of a word using boundary condition and included plurals and 's..

* Missing/No Features  Found- Concept --- Improved concept by using synonyms and hyponyms and blocking out the sentence.

* Small amount of Features Found --- Addresses Improved address by using a better regex pattern.

* Missing Stats --- Fixed stats stderr, stdout.

* Missing Test Functions --- Added tests

* Errors Running Test --- Tests working from repo folder. Had to change from project1 import redactor to just import redactor.

* was not preserving the format of the file in input. Fixed that.


tests/test_redactor.py
import redactor
import pytest
import nltk
path = './tests/5.txt'
def test_datainput ():
    data = redactor.input_files(path)
    print (len(data))
    if len(data) != 0:
        assert data is not None



def test_phonenumbers():
    data = redactor.input_files(path)
    phonenumbers = redactor.redact_all_phonenumbers(data)    
    assert len(phonenumbers) is not None
   

def test_genders():
    count =0
    data = (redactor.input_files(path))
    gender_lst = redactor.redact_all_genders(data)
    for i in range(len(gender_lst)):
        temp = gender_lst[i]
        words = nltk.word_tokenize(temp)
        for j in words:
            if '\u2588' in j:
                count += 1
    assert count is not None

def test_dates():

    count = 0

    data= redactor.input_files(path)
    redacted_data  = redactor.redact_all_dates(data)
    for i in range(len(redacted_data)):
        redaction_style = u"\u2588"
        for j in range(len(redacted_data[i])):
            word = redacted_data[i][j]
            redaction_style = len(word)*redaction_style
            if redaction_style == word:
                count = count + 1
    if count >= 0 or len(redacted_data) > 0 :
        assert True


def test_names():
    
    count =0
    data = (redactor.input_files(path))
    name_lst = redactor.redact_all_names(data)
    for i in range(len(name_lst)):
        temp = name_lst[i]
        names = nltk.word_tokenize(temp)
        for j in names:
            if '\u2588' in j:
                count += 1
    assert count is not None


tests/testing_redactor.py
from project1 import redactor
import pytest
import nltk
path = './tests/5.txt'
def test_datainput ():
    data = redactor.input_files(path)
    print (len(data))
    if len(data) != 0:
        assert data is not None



def test_phonenumbers():
    data = redactor.input_files(path)
    phonenumbers = redactor.redact_all_phonenumbers(data)    
    assert len(phonenumbers) is not None
   

def test_genders():
    count =0
    data = (redactor.input_files(path))
    gender_lst = redactor.redact_all_genders(data)
    for i in range(len(gender_lst)):
        temp = gender_lst[i]
        words = nltk.word_tokenize(temp)
        for j in words:
            if '\u2588' in j:
                count += 1
    assert count is not None

def test_dates():

    count = 0

    data= redactor.input_files(path)
    redacted_data  = redactor.redact_all_dates(data)
    for i in range(len(redacted_data)):
        redaction_style = u"\u2588"
        for j in range(len(redacted_data[i])):
            word = redacted_data[i][j]
            redaction_style = len(word)*redaction_style
            if redaction_style == word:
                count = count + 1
    if count >= 0 or len(redacted_data) > 0 :
        assert True


def test_names():
    
    count =0
    data = (redactor.input_files(path))
    name_lst = redactor.redact_all_names(data)
    for i in range(len(name_lst)):
        temp = name_lst[i]
        names = nltk.word_tokenize(temp)
        for j in names:
            if '\u2588' in j:
                count += 1
    assert count is not None


project1/main.py
import argparse
import redactor


if __name__ == "__main__":
    

    arguments = argparse.ArgumentParser()
    arguments.add_argument("--input",type = str, required = True, help = "path to input files", nargs = "*", action = "append" )
    arguments.add_argument("--names", required = False, help = "option to redate names", action = "store_true")
    arguments.add_argument("--genders", required = False, help = "option to redact genders", action = "store_true")
    arguments.add_argument("--dates", required = False, help = "option to redact dates", action = "store_true")
    arguments.add_argument("--phones", required = False, help = "option to redact phone numbers", action = "store_true")
    arguments.add_argument("--phones", required = False, help = "option to redact phone numbers", action = "store_true")
    # can be repeated multiple times
    arguments.add_argument("--concept", type = str, required = False, help = "option to redact concepts", nargs= "*", action = "append")
    arguments.add_argument("--stats", type = str, required = False, help = "option to to print stats for redactions")
    arguments.add_argument("--output", type = str, required = True, help = "path  to outputfile")

    args = arguments.parse_args()
    data  = redactor.input_files(args.input)

    if args.names:
        data = redactor.redact__all_names(data)
    
    if args.dates:
        data = redactor.redact_all_dates(data)
    
    if args.phones:
        data = redactor.redact_all_phonenumbers(data)
        

    if args.genders:
        data = redactor.redact_all_genders(data)

    if args.concept:
        data = redactor.redact_all_concepts(data, args.concept)
        
    if args.output:
        redactor.get_output(args.input, data, args.output)

    if args.stats:
        
        redactor.write_stats()








redactor.py
import argparse
import os
import re 
import spacy
from spacy import displacy
import nltk
import glob
from commonregex import CommonRegex
import sys
# named entity recognition using spacy
import en_core_web_md
nlp = en_core_web_md.load()
#nlp = spacy.load("en_core_web_sm")
nltk.download('wordnet')
nltk.download('omw-1.4')
#nltk.download('words')
#nltk.download('wordnet')
#nltk.download('punkt')
#nltk.download('maxent_ne_chunker')
#nltk.download('averaged_perceptron_tagger')
#nltk.download('all')
from itertools import chain

from nltk.corpus import wordnet 
from nltk.tokenize import word_tokenize,sent_tokenize


# the user can input multiple files
def input_files(files_in):
    '''
    if the target token is next to \n or there special characters it leads to incorrect label assignment there for removing \n and other special characters
    '''
    if len(files_in) == 0:
        raise Exception('No files given\n')
    files_together = []
    all_data = ""
    # make nested list into list
    files_in = nltk.flatten(files_in)
    for text_files in files_in:

            lst_fls = glob.glob(text_files)

            for each_file in lst_fls:
                data = open(each_file,"r").read()
                #data = data.replace("\\","")
                data = data + "!@#$%^&*("
                all_data = ''.join([all_data,data])
             #   data = data + "&&%#&*&"
               # data = data.replace("\n"," ")
               # data = data.replace("_"," ")
                
                #all_data.append("\t+")
   
    
   # print(all_data)
    return all_data
def redact_all_names(data):
    '''
    does not redact uncommon names like Ermis 
    does not redact name if it appears in the email address
    Eventhough it redacts common names like Frank spacy incorrectly assigns label PERSON to ['Frank Ermis Jan2002', 'Frank Inbox X-Origin']
    '''
    type_of = "Names:"
    doc = nlp(data)
    names = []
    for ent in doc:
       # print(ent.text)
        if ent.ent_type_ =='PERSON':
            names.append(ent.text) 
  #  names =  [ent.text for ent in doc.ents if ent.label_ == 'PERSON' ]
    #print(names)   
    for name in names:
           
         data = re.sub(name,'\u2588' * len(name), data)
    count = len(names)
   # print(names,count)
    get_stats(type_of, count)
   # print(data)
    return data

 
def redact_all_dates(data):
    '''
    correctly matches dates like this 25 Nov 2001
    incorrectly matches like ['26 Nov 0001', '4-20-7783', '1-2-9229'] sections phone numbers but does not redact them because it matches a subsection of the token
    '''
    type_of = "Dates:"
    
    redacted_dates = CommonRegex(data).dates
    
    for date in redacted_dates:
        
         data = re.sub('\\b{}\\b'.format(date),'\u2588' * len(date), data)
    count = len(redacted_dates)
   # print(redacted_dates)
   # print(data)
    get_stats(type_of,count)
    return data

def redact_all_phonenumbers(data):
    '''
    redacts wrong 2 numbers at the start ['5941235', '1075855165188', '(713) 345-4727']
    won't redact phones of the format +61-2-9229-2336
    correctly matches phonenubmers of the format (713) 345-4727 
    '''
    type_of = "Phones:"
    
    redacted_numbers = CommonRegex(data).phones
   # print(redacted_numbers)   
    for number in redacted_numbers:
        data = data.replace(number,'\u2588' * len(number))
   # print(data)
    count = len(redacted_numbers) 
    get_stats(type_of,count)
   # print(data)
    return data
    

def redact_all_genders(data):
    type_of = "Genders:"
    gender_indicators = ['HE','HIM','SHE','HER','FATHER','MOTHER','MOM','DAD','MUMMY','DADDY', 'BROTHER','SISTER','UNCLE','AUNT', 'BOY', 'GIRL', 'BOYFRIEND','GIRLFRIEND','POP','MEN','WOMEN','ACTOR','ACTRESS','NIECE','NEPHEW','wife','husband','boy','girl','man','woman','he','she','daughter','son','niece','nephew','aunt','mother','male','female','girlfriend','boyfriend','gentlemen','gentlewoman','brother','sister','grandmother','grandfather','granddaughter','bride','groom','sir','grandma','grandpa','grandmom','herself','himself','queen','king','princess','prince','women','widower','widow','fiancee','fiance','herione','hero','herione','god','goddess','lord','lady','chairman','chairwoman'] 
   # tokens = nltk.word_tokenize(data)
   
  #  count = 0
  #  for word in tokens:   
  #     if word.upper() in gender_indicators:
  #          data = data.replace(word,"\u2588" * len(word))
  #          count = count + 1
    
  #  get_stats(type_of,count)
   # print(data)i
#    reg = re.compile(r"\b(?:(" + "(')?s?)|(".join(gender_indicators) + r"))\b", flags=re.I)
    
    counts = 0
    for gender in gender_indicators:
       

            data,c = re.subn("\\b{}\\b|\\b{}s\\b|\\b{}'s\\b".format(gender,gender,gender) , '\u2588'*len(gender), data, flags = re.I,count =0)
            counts = counts + c
    get_stats(type_of,counts)

    
    return data


def redact_all_addressess(data):
   
    type_of ="Addressess:"
   # pattern = '(\d{1,6}[A-Za-z]?\s(?:[A-Za-z0-9#-]+\s){0,7}(?:[A-Za-z0-9#-]+,)\s(?:[A-Za-z]+\s){0,3}(?:[A-Za-z]+,)\s[A-Z]{2}\s*\d{4,5})'
# addressess =re.findall(r'\d{1,6}\s(?:[A-Za-z0-9#]+\s){0,7}(?:[A-Za-z0-9#]+,)\s*(?:[A-Za-z]+\s){0,3}(?:[A-Za-z]+,)\s*[A-Z]{2}\s*\d{5}' ,data)
  #  addressess =re.findall('(\d{1,6}[A-Za-z]?\s(?:[A-Za-z0-9#]+\s){0,7}(?:[A-Za-z0-9#]+,)\s*(?:[A-Za-z]+\s){0,3}(?:[A-Za-z]+,)\s([A-Z]{2})\s(\d{4,5})*)' ,data)
    # findall gives empty str for no match   
    addressess =re.findall('(\d{1,6}[A-Za-z]?\s(?:[A-Za-z0-9#]+\s){0,7}(?:[A-Za-z0-9#]+,)\s*(?:[A-Za-z]+\s){0,3}(?:[A-Za-z]+,*)\s([A-Z]{2})*\s*\d{4,5})' ,data)
 
    
   # if len(addressess) > 0:
   # print(addressess)   
    for address in addressess:
    
        if len(address) > 0:
               
            i = '\u2588' * len(address[0])
        
            data = re.sub(address[0],i, data )
  # print(data)
    count = len(addressess)
    get_stats(type_of,count)
    return data
     

def redact_all_concepts(data,concept):
    '''
    repace every new line with .
    '''
   # doc = nlp(data)
   # for sentence in doc.sents:
   #     print("INSIDE for")
   #     print(sentence.text)
    con =  concept[0][0] 
    synonyms = [[con]]
    type_of = "Concepts:"
    counts = 0

    x=[]
    for hypo in  wordnet.synsets(con)[0].hyponyms():
        synonyms.append(hypo.lemma_names())

    for syn in wordnet.synsets(con):
        for l in syn.lemmas():
            x.append(l.name())
        synonyms.append(x)    

    synonyms = list(chain.from_iterable(synonyms))
    data1 = data.replace("\n",". ")   
    sentences = data1.split(". ")
    sentences = list(filter(None,sentences))
    
    for sent in sentences:
        for con in synonyms:
            match = re.findall(con,sent,flags = re.I)
           
            if len(match) > 0:
                replace = '\u2588'* len(sent)
                data,c = re.subn(sent, replace ,data,count = 0)
                counts = counts + c
   # print(counts)
    get_stats(type_of,counts)    
    for syn in synonyms:

        data = re.sub(syn,'\u2588' * len(syn),data )

    return data
def redacted_output(files_in,data,redacted_file):
   # print("BEFORE SPLIT\n\n\n",data) 
    data = data.split("!@#$%^&*(")
   # print(len(data))
  
    lst_files = []
    for i in files_in:
        for file in i:
            lst_files.append(glob.glob(file))
    flat = nltk.flatten(lst_files)
   # print(flat)
    # save redacted file in current working directory
    files_at = os.path.join(os.getcwd(), redacted_file)
    # redacted_file = file name that the user inputs
   #files_at =  home/callpradip96/cs5293sp22-project1/project1/t
   # print(files_at)
    # replace imput file extension with .redacted
    filen = 0
   # print(len(flat))
    while filen < len(flat):
      
       # ipath = os.path.splitext(flat[filen])[0]
        ipath = os.path.basename(flat[filen]) + ".redacted"
   
        if not os.path.exists(files_at):
            os.makedirs(files_at)
            with open(os.path.join(files_at,ipath), 'w') as output:
      #          print(data[filen])
                output.write(data[filen])
    
        elif os.path.exists(files_at):
            with open(os.path.join(files_at, ipath), 'w') as output:
              #  print(data[filen])
                output.write(data[filen])
     #   print("BEFORE",filen)
        filen += 1
    
store_stats = []
def get_stats(type_of,count):
    count = str(count)
    if type_of == "Names:":
        
        store_stats.append(type_of + str(count))
     
    if type_of == "Dates:":
        store_stats.append(type_of + str(count))
        

    if type_of == "Phones:":
        store_stats.append(type_of + str(count))

    if type_of == "Genders:":
        store_stats.append(type_of + str(count))

    if type_of == "Addressess:":
        store_stats.append(type_of + str(count))

    if type_of == "Concepts:":
        store_stats.append(type_of + str(count))

    
    return store_stats



def write_stats(pathd,store_stats):
    '''
    Creates a file in the current working directory and writes the stats to the file
    '''
  #  print(store_stats)   

    # create a folder and write a file inside it
   

    if(pathd != 'stdout' and pathd != 'stderr' ):
    
    
        with open(pathd,'w') as statout:
            for stat in store_stats:
                statout.write(stat)
                statout.write('\n')
        statout.close()
    else:
        for stat in store_stats:
            if(pathd == 'stdout'):
                sys.stdout.write(stat)      
                sys.stdout.write('\n')
            if(pathd == 'stderr'):
                sys.stderr.write(stat)
                sys.stderr.write('\n')
    
if __name__ == "__main__":

    arguments = argparse.ArgumentParser()
    arguments.add_argument("--input",type = str, required = True, help = "path to input files", nargs = "*", action = "append" )
    arguments.add_argument("--names", required = False, help = "option to redate names", action = "store_true")
    arguments.add_argument("--genders", required = False, help = "option to redact genders", action = "store_true")
    arguments.add_argument("--dates", required = False, help = "option to redact dates", action = "store_true")
    arguments.add_argument("--phones", required = False, help = "option to redact phone numbers", action = "store_true")
    arguments.add_argument("--address", required = False, help = "option to redact address", action = "store_true")
    # can be repeated multiple times
    arguments.add_argument("--concept", type = str, required = False, help = "option to redact concepts", nargs= "*", action = "append")
    arguments.add_argument("--stats", type = str, required = False, help = "option to to print stats for redactions")
    arguments.add_argument("--output", type = str, required = True, help = "path  to outputfile")
    args = arguments.parse_args()

    data  = input_files(args.input)

    if args.concept:
        data = redact_all_concepts(data, args.concept)
  
    if args.phones:
        data = redact_all_phonenumbers(data)
    
    if args.address:
        data = redact_all_addressess(data)

    if args.dates:
        data = redact_all_dates(data)

    if args.genders:
        data = redact_all_genders(data)

    if args.names:
        data = redact_all_names(data)
                     
    if args.stats:
        write_stats(args.stats,store_stats)
    if args.output:
        redacted_output(args.input, data, args.output)

